### ISSUE TITLE
@import CSS files are not being updated

### ISSUE THREAD

**Issue by processprocess** [ampproject/amphtml#33826]
imported CSS files are not being updated in the build.

I'm noticing this with the imports in `amp-story.css`.

To repro:
1. Start dev server
2. Navigate to a story example on desktop.
3. Make an edit in `amp-story-desktop-panels.css`
4. Refresh page, edits are not visible.

Sometimes works when following these steps:
1. make edit to amp-story-draggable-drawer.css
2. make edit to amp-story.css
3. refresh sample in local server
4. undo step 2

**Comment by rsimha** [comment]
My suspicion is that is due to #33313, which introduced caching for CSS. I wonder if cache invalidation is not working as expected. Assigning to @samouri for further investigation.

**Comment by alanorozco** [comment]
The cache is being keyed by the contents of the `.css` entry point, not by the contents of its references.

If it were to check the contents of its `@import` references, the cache would be properly invalidated.

The following changes to `build-system/tasks/css/jsify-css.js` fix this problem. I'll leave it to @samouri for the actual fix.

```diff
  async function jsifyCssAsync(filename) {
-   const {contents, hash: filehash} = await batchedRead(filename);
+   const {contents, hash: filehash} = await referencedBatchedRead(filename);

    const hash = md5(filehash, await getEnvironmentHash());
    const result = await transformCss(contents, hash, filename);

    result.warnings.forEach((warn) => log(red(warn)));
    return result.css + '\n/*# sourceURL=/' + filename + '*/';
  }

+ async function referencedBatchedRead(filename) {
+   const {contents, hash} = await batchedRead(filename);
+   const hashes = [hash];
+   for (const referencedFilename of collectCssFilenames(
+     path.dirname(filename),
+     contents
+   )) {
+     const {hash} = await batchedRead(referencedFilename);
+     hashes.push(hash);
+   }
+   return {contents, hash: md5(...hashes)};
+ }

+ function* collectCssFilenames(dirname, contents) {
+   const regex = /['"]([^'"]+\.css)['"]/g;
+   let match;

+   while ((match = regex.exec(contents)) !== null) {
+     // eslint-disable-next-line no-unused-vars
+     const [unusedFullMatch, filepath] = match;
+     yield path.resolve(dirname, filepath);
+   }
+ }
```

**Comment by samouri** [comment]
@alanorozco: please feel free to make a PR! I unfortunately don't have the bandwidth for this... so my fix would be to disable CSS caching for any file that imports others. This looks much better

**Comment by alanorozco** [comment]
I also lack the bandwidth to land this, hence my quick-and-dirty patch. @rsimha? 😬

**Comment by alanorozco** [comment]
@rsimha Have you taken this on? It will be fixed collaterally as part of #34274

**Comment by raxsha** [comment]
Just as an FYI, this is also blocking a fix for #34464

**Comment by kristoferbaxter** [comment]
@raxsha how is this blocking a fix? A restart of the CLI should allow everything to work as expected. (And even with a restart this is faster than the before state.)

**Comment by processprocess** [comment]
I noticed that changes in `extensions/amp-story/1.0/amp-story-desktop-one-panel.css` require a restart even with the `jsify-css.js` patch.  

It uses an entrypoint here:
https://github.com/processprocess/amphtml/blob/main/build-system/tasks/css/index.js#L72

It's injected into shadow dom on the document here:
https://github.com/ampproject/amphtml/blob/main/src/amp-story-player/amp-story-player-impl.js#L38

Restarting unblocks me but making a note here since it might be related.

**Comment by gmajoulet** [comment]
Friendly ping @ampproject/wg-infra
If this is not on your roadmap at all, please consider lowering the priority of this ticket so we look for alternative solutions.

**Comment by rsimha** [comment]
> If this is not on your roadmap at all, please consider lowering the priority of this ticket so we look for alternative solutions.

For context, detection of changes to transitive CSS dependencies was never an advertised feature of the `amp` lazy builder, so I've changed this issue to P2 / Feature request instead of P1 / Bug (as initially filed by Philip). In addition, infra's main focus since mid 2021 has been open-sourcing release / serving infra and Bento dev-ex, as a result of which new FRs to the dev toolchain have remained on the back-burner.

> What's the recommended workaround for this?

As described in https://github.com/ampproject/amphtml/issues/33826#issuecomment-836938979, the ideal solution to this problem is being pursued in https://github.com/ampproject/amphtml/pull/34274. If that solution can't be achieved for any reason, a less ideal workaround may be possible. I've looped in @alanorozco for comment.
